### PR TITLE
feat: optional entry.{client,server} for react 17

### PR DESCRIPTION
**Pull request by mcansh**
Closes: #5669
Closes: #5810

- [ ] Docs
- [x] Tests

Testing Strategy:

### CODE REVIEW

**Review by MichaelDeBoey — APPROVED** [review]

File: packages/remix-dev/config.ts [hunk 1]
```diff
@@ -726,11 +737,18 @@ const resolveServerBuildPath = (
 
 // @ts-expect-error available in node 12+
 // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/ListFormat#browser_compatibility
-let listFormat = new Intl.ListFormat("en", {
+let conjunctionListFormat = new Intl.ListFormat("en", {
```
Comment (jacob-ebey) [inline packages/remix-dev/config.ts:740]: What is this magic doing?

Comment (mcansh) [reply packages/remix-dev/config.ts:740]: basically just converts an array to "x and y", "x, y, and z", "x or y", "x, y, or z"

File: packages/remix-dev/config.ts [hunk 2]
```diff
@@ -463,7 +464,49 @@ export async function readConfig(
...
+      throw new Error(
+        `Could not determine server runtime. Please install one of the following: ${formattedList}`
+      );
+    }
+
+    let clientRuntime = deps["@remix-run/react"] ? "react" : undefined;
+
+    if (!clientRuntime) {
+      throw new Error(
+        `Could not determine runtime. Please install the following: @remix-run/react`
```
Comment (MichaelDeBoey) [inline packages/remix-dev/config.ts:491]:
```suggestion
        `Could not determine renderer. Please install the following: @remix-run/react`
```

File: packages/remix-dev/config.ts [hunk 3]
```diff
@@ -485,27 +528,7 @@ export async function readConfig(
...
-        "@remix-run/node",
-      ];
-      let formattedList = listFormat.format(serverRuntimes);
-      throw new Error(
-        `Could not determine server runtime. Please install one of the following: ${formattedList}`
-      );
-    }
-
-    entryServerFile = `entry.server.${serverRuntime}.tsx`;
+    entryServerFile = `${serverRuntime}/entry.server.${clientRuntime}-${type}.tsx`;
```
Comment (MichaelDeBoey) [inline packages/remix-dev/config.ts:531]:
```suggestion
    entryServerFile = `${serverRuntime}/entry.server.${clientRenderer}-${type}.tsx`;
```

File: packages/remix-dev/config.ts [hunk 4]
```diff
@@ -519,7 +542,21 @@ export async function readConfig(
...
+        `Could not determine React version. Please install the following packages: ${list}`
+      );
+    }
+
+    let type: "stream" | "string" =
+      maybeReactVersion.major >= 18 || maybeReactVersion.raw === "0.0.0"
+        ? "stream"
+        : "string";
+
+    entryClientFile = `entry.client.${clientRuntime}-${type}.tsx`;
```
Comment (MichaelDeBoey) [inline packages/remix-dev/config.ts:559]:
```suggestion
    entryClientFile = `entry.client.${clientRenderer}-${type}.tsx`;
```

File: packages/remix-dev/cli/commands.ts [hunk 5]
```diff
@@ -326,11 +348,12 @@ export async function generateEntry(
   let defaultsDirectory = path.resolve(__dirname, "..", "config", "defaults");
   let defaultEntryClient = path.resolve(
     defaultsDirectory,
-    `entry.client.${clientRuntime}.tsx`
+    `entry.client.${clientRuntime}-${type}.tsx`
```
Comment (MichaelDeBoey) [inline packages/remix-dev/cli/commands.ts:351]:
```suggestion
    `entry.client.${clientRenderer}-${type}.tsx`
```

File: packages/remix-dev/cli/commands.ts [hunk 6]
```diff
@@ -326,11 +348,12 @@ export async function generateEntry(
...
   );
   let defaultEntryServer = path.resolve(
     defaultsDirectory,
-    `entry.server.${serverRuntime}.tsx`
+    serverRuntime,
+    `entry.server.${clientRuntime}-${type}.tsx`
```
Comment (MichaelDeBoey) [inline packages/remix-dev/cli/commands.ts:356]:
```suggestion
    `entry.server.${clientRenderer}-${type}.tsx`
```

**Review by MichaelDeBoey — APPROVED** [review]

**Review by MichaelDeBoey — APPROVED** [review]

**Comment by ZackWard** [comment]
@mcansh I just discovered this change (optional entry files) today and found it pretty confusing. I would like to propose that along with this change, the prominent `docs/file-conventions/entry.client.md` and `docs/file-conventions/entry.server.md` files be updated to clarify that users should not expect to see these files by default.

**Review by MichaelDeBoey — APPROVED** [review]

**Pull request merged**